provision/provision_keybox.cpp

```diff
@@ -153,54 +148,184 @@ static uint8_t *decompress_attkb(uint8_t* keybox, size_t compressed_size) {
+    if (!kb_data) {
+        LOG_E("Failed to alloc buf for keybox:%d.\n", encrypt_attkb_size);
+        storage_close_session(session);
+        return KM_ERROR_MEMORY_ALLOCATION_FAILED;
+    }
+    memset(kb_data, 0, encrypt_attkb_size);
+
+    res = storage_read_attkb(session, kb_data, encrypt_attkb_size);
+    if ((size_t)res != encrypt_attkb_size) {
+        LOG_E("Failed to read keybox from SS, res = %d\n", res);
+        free(kb_data);
+        storage_close_session(session);
+        return KM_ERROR_UNKNOWN_ERROR;
+    }
+
+    storage_close_session(session);
+
+    *keybox_size = encrypt_attkb_size;
+    *keybox = (uint8_t *)kb_data;
+
+    return KM_ERROR_OK;
+}
+
+#else
+/* Keybox is retrieved from the CSE directly. */
+static keymaster_error_t get_keybox_from_cse(trusty_device_info_t * dev_info, uint8_t** keybox, uint32_t* keybox_size) {
+    *keybox_size = dev_info->attkb_size;
+    if (*keybox_size == 0) {
+        LOG_E("Failed to get keybox from CSE.\n", 0);
+        return KM_ERROR_UNKNOWN_ERROR;
+    }
+
+    *keybox = (uint8_t *)malloc(*keybox_size);
+    if (*keybox == NULL) {
+        return KM_ERROR_UNKNOWN_ERROR;
+    }
+
+    memcpy_s(*keybox, *keybox_size, dev_info->attkb, *keybox_size);
+
+    return KM_ERROR_OK;
+}
+#endif
+
+static keymaster_error_t decrypt_attkb(trusty_encrypted_attkb_t *attkb, uint8_t *attkb_enc_key,
+                                   uint8_t* keybox, uint32_t* keybox_size, size_t *out_size) {
     int rc = -1;
+    trusty_encrypted_attkb_t aad;
+
+    /* AAD for GCM is struct trusty_encrypted_attkb_t with tag cleared */
+    memcpy_s(&aad, sizeof(trusty_encrypted_attkb_t), attkb, sizeof(trusty_encrypted_attkb_t));
+    memset(aad.trusty_cipher_blob.tag, 0, sizeof(attkb->trusty_cipher_blob.tag));
+
+    size_t cipher_len = attkb->trusty_cipher_blob.cipher_blob_size + sizeof(attkb->trusty_cipher_blob.tag);
+    void *cipher = (void *)malloc(cipher_len);
+    if (cipher == NULL) {
+        return KM_ERROR_MEMORY_ALLOCATION_FAILED;
+    }
+
+    /* exchange cipher and tag sequence to match aes_256_gcm_decrypt API */
+    memcpy_s(cipher, attkb->trusty_cipher_blob.cipher_blob_size,
+             attkb->trusty_cipher_blob.cipher_blob, attkb->trusty_cipher_blob.cipher_blob_size);
+    memcpy_s((uint8_t*)cipher + attkb->trusty_cipher_blob.cipher_blob_size, sizeof(attkb->trusty_cipher_blob.tag),
+             attkb->trusty_cipher_blob.tag, sizeof(attkb->trusty_cipher_blob.tag));
+
+    rc = aes_256_gcm_decrypt((const struct gcm_key *)attkb_enc_key,
+                             (const void *)attkb->trusty_cipher_blob.iv,
+                             sizeof(attkb->trusty_cipher_blob.iv),
+                             (const void *)&aad,
+                             sizeof(trusty_encrypted_attkb_t),
+                             (const void *)cipher,
+                             cipher_len,
+                             keybox + sizeof(attkb_header_t), out_size);
+    free(cipher);
```

Comment:
does cipher contain secrets? if yes, better zero it before free()

Reply:
No.  But it's a good practice to zero it.